### PR TITLE
[SYCL][ESIMD] add indirect read and write methods

### PR DESCRIPTION
**Pull request by cmc-rep**
Signed-off-by: Gang Y Chen <gang.y.chen@intel.com>

### CODE REVIEW

**Comment by cmc-rep** [comment]
@kychendev @kbobrovs please review and merge timely

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_intrin.hpp [hunk 1]
```diff
@@ -261,6 +273,19 @@ __esimd_rdregion(sycl::INTEL::gpu::vector_type_t<T, N> Input, uint16_t Offset) {
...
 }
 
+template <typename T, int N, int M, int ParentWidth>
+SYCL_EXTERNAL sycl::INTEL::gpu::vector_type_t<T, M>
+__esimd_rdindirect(sycl::INTEL::gpu::vector_type_t<T, N> Input,
+                   sycl::INTEL::gpu::vector_type_t<uint16_t, M> Offset) {
+  sycl::INTEL::gpu::vector_type_t<T, M> Result;
+  for (int i = 0; i < M; ++i) {
+    uint16_t EltOffset = Offset[i] / sizeof(T);
+    assert(Offset[i] % sizeof(T) == 0);
```
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_intrin.hpp:283]: Could add safety check here and report error on out-of-bound access.

File: sycl/include/CL/sycl/INTEL/esimd/detail/esimd_intrin.hpp [hunk 2]
```diff
@@ -285,4 +310,21 @@ __esimd_wrregion(sycl::INTEL::gpu::vector_type_t<T, N> OldVal,
...
+SYCL_EXTERNAL sycl::INTEL::gpu::vector_type_t<T, N>
+__esimd_wrindirect(sycl::INTEL::gpu::vector_type_t<T, N> OldVal,
+                   sycl::INTEL::gpu::vector_type_t<T, M> NewVal,
+                   sycl::INTEL::gpu::vector_type_t<uint16_t, M> Offset,
+                   sycl::INTEL::gpu::mask_type_t<M> Mask) {
+  sycl::INTEL::gpu::vector_type_t<T, N> Result = OldVal;
+  for (int i = 0; i < M; ++i) {
+    if (Mask[i]) {
+      uint16_t EltOffset = Offset[i] / sizeof(T);
+      assert(Offset[i] % sizeof(T) == 0);
```
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/detail/esimd_intrin.hpp:323]: Could add range check here too.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 3]
```diff
@@ -179,9 +179,29 @@ template <typename Ty, int N> class simd {
   //   This would allow you to use the subscript operator to write to an
   //   element.
   // {/quote}
-  /// Read a single element, by value only.
+  /// Read single element, return value only (not reference).
   Ty operator[](int i) const { return data()[i]; }
 
+  /// Read multiple elements by their indices in vector
+  template <int Size> simd<Ty, Size> iselect(simd<uint16_t, Size> &Indices) {
```
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:186]: Could add "const" for Indices parameter.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 4]
```diff
@@ -179,9 +179,29 @@ template <typename Ty, int N> class simd {
...
+  }
+  /// update single element
+  void iupdate(ushort Index, Ty V) {
+    auto Val = data();
+    Val[Index] = V;
+    set(Val);
+  }
+  /// update multiple elements by their indices in vector
+  template <int Size>
+  void iupdate(simd<uint16_t, Size> &Indices, simd<Ty, Size> &Val,
```
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:198]: Could add "const" for Indices and Val parameters.

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 5]
```diff
@@ -179,9 +179,29 @@ template <typename Ty, int N> class simd {
...
+               mask_type_t<Size> Mask) {
+    vector_type_t<uint16_t, Size> Offsets = Indices.data() * sizeof(Ty);
+    M_data = __esimd_wrindirect<Ty, N, Size>(data(), Val.data(), Offsets, Mask);
+    set(M_data);
```
Comment (kychendev) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:202]: This is redundant for setting M_data? Or do we need a temp to perform v_store to M_data?

**Comment by cmc-rep** [comment]
@kychendev @kbobrovs , update is ready for review

**Review by kychendev — APPROVED** [review]

**Comment by cmc-rep** [comment]
@konst, could you help with this merge?

**Comment by cmc-rep** [comment]
@kbobrovs, could you help with this merge

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**